glean-core/tests/clientid_textfile.rs

```diff
@@ -0,0 +1,314 @@
+        disabled: false,
+        dynamic_label: None,
+    })
+}
+
+fn clientid_from_file(data_path: &Path) -> Option<Uuid> {
+    let path = data_path.join("client_id.txt");
+    let uuid_str = fs::read_to_string(path).ok()?;
+    Uuid::parse_str(uuid_str.trim_end()).ok()
+}
+
+fn write_clientid_to_file(data_path: &Path, uuid: &Uuid) -> Option<()> {
+    let path = data_path.join("client_id.txt");
+    let mut buffer = Uuid::encode_buffer();
+    let uuid_str = uuid.hyphenated().encode_lower(&mut buffer);
+    fs::write(path, uuid_str.as_bytes()).ok()?;
+    Some(())
+}
+
+#[test]
+fn writes_clientid_file() {
+    let (glean, temp) = new_glean(None);
+
+    let db_client_id = clientid_metric().get_value(&glean, None).unwrap();
+    let file_client_id = clientid_from_file(temp.path()).unwrap();
+
+    assert_eq!(file_client_id, db_client_id);
+
+    glean.submit_ping_by_name("health", Some("pre_init"));
+    let mut pending = get_queued_pings(temp.path()).unwrap();
+    assert_eq!(1, pending.len());
+    let payload = pending.pop().unwrap().1;
+
+    let state = &payload["metrics"]["string"]["glean.health.exception_state"];
+    assert_eq!(&serde_json::Value::Null, state);
+}
+
+#[test]
+fn reused_clientid_from_file() {
+    let temp = tempfile::tempdir().unwrap();
+    let new_uuid = Uuid::new_v4();
+
+    write_clientid_to_file(temp.path(), &new_uuid).unwrap();
+
+    let (glean, temp) = new_glean(Some(temp));
+    // We don't run the mitigation yet
+    //let db_client_id = clientid_metric().get_value(&glean, None).unwrap();
+    //assert_eq!(new_uuid, db_client_id);
+
+    glean.submit_ping_by_name("health", Some("pre_init"));
+    let mut pending = get_queued_pings(temp.path()).unwrap();
+    assert_eq!(1, pending.len());
+    let payload = pending.pop().unwrap().1;
+
+    let state = payload["metrics"]["string"]["glean.health.exception_state"].as_str();
+    assert_eq!(Some("empty-db"), state);
+}
+
+#[test]
+fn restores_clientid_file_from_db() {
+    let (db_client_id, temp) = {
+        // Ensure we initialize once to get a client_id
+        let (glean, temp) = new_glean(None);
+        let db_client_id = clientid_metric().get_value(&glean, None).unwrap();
+        drop(glean);
+
+        (db_client_id, temp)
+    };
+
+    // Removing the file. `Glean::new` should restore it
+    fs::remove_file(temp.path().join("client_id.txt")).unwrap();
+
+    let (glean, temp) = new_glean(Some(temp));
+    let file_client_id = clientid_from_file(temp.path()).unwrap();
+    assert_eq!(file_client_id, db_client_id);
+
+    let db_client_id2 = clientid_metric().get_value(&glean, None).unwrap();
+    assert_eq!(db_client_id, db_client_id2);
+
+    glean.submit_ping_by_name("health", Some("pre_init"));
+    let mut pending = get_queued_pings(temp.path()).unwrap();
+    assert_eq!(1, pending.len());
+    let payload = pending.pop().unwrap().1;
+
+    let state = payload["metrics"]["string"]["glean.health.exception_state"].as_str();
```

Comment:
musing: Should we consider an absent file to be a state worth reporting... I mean, obviously, the first init after updating will always fall into this case. But if it happens later, would we want to know? Would we want to monitor this over time?